org.sonarlint.eclipse.m2e/src/org/sonarlint/eclipse/m2e/internal/M2eUtils.java

```diff
@@ -0,0 +1,110 @@
+import org.eclipse.core.internal.resources.ICoreConstants;
+import org.eclipse.core.internal.resources.Resource;
+import org.eclipse.core.resources.IContainer;
+import org.eclipse.core.resources.IFile;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.resources.ResourcesPlugin;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.m2e.core.MavenPlugin;
+import org.eclipse.m2e.core.internal.IMavenConstants;
+import org.eclipse.m2e.core.project.IMavenProjectFacade;
+import org.eclipse.m2e.core.project.IMavenProjectRegistry;
+import org.sonarlint.eclipse.core.SonarLintLogger;
+
+public class M2eUtils {
+
+  private M2eUtils() {
+  }
+
+  /**
+   * m2e creates IProject for every module (flat layout). It means the same file can appear in several projects.
+   * When it is the case we want to keep the most specific one.
+   */
+  public static boolean isInNestedModule(IFile file) {
+    IProject project = file.getProject();
+    try {
+      if (project.hasNature(IMavenConstants.NATURE_ID)) {
+        IMavenProjectRegistry projectManager = MavenPlugin.getMavenProjectRegistry();
+
+        IMavenProjectFacade projectFacade = projectManager.create(project, null);
+        if (projectFacade != null && "pom".equals(projectFacade.getPackaging())) {
+          return !toSpecificFile(file).equals(file);
+        }
+      }
+    } catch (CoreException ex) {
+      SonarLintLogger.get().error(ex.getMessage(), ex);
+    }
+    return false;
+  }
+
+  private static IFile toSpecificFile(IFile file) {
+    IFile finalFile = file;
+    IPath rawLocation = file.getRawLocation();
+    if (rawLocation != null) {
+      // TODO use getFileForLocation when we will support only Eclipse 4.6+
+      // IFile moreSpecific = ResourcesPlugin.getWorkspace().getRoot().getFileForLocation(rawLocation);
+      IFile moreSpecific = resourceForLocation(rawLocation);
+      if (moreSpecific != null) {
+        finalFile = moreSpecific;
+      }
+    }
+    return finalFile;
+  }
+
+  /**
+   * Copied from {@link FileSystemResourceManager} of Oxygen to support older Eclipse versions
+   */
+  private static IFile resourceForLocation(IPath location) {
+    int resultProjectPathSegments = 0;
+    IFile result = null;
+    IProject[] projects = ResourcesPlugin.getWorkspace().getRoot().getProjects(IContainer.INCLUDE_HIDDEN);
+    for (int i = 0; i < projects.length; i++) {
+      IProject project = projects[i];
+      IPath projectLocation = project.getLocation();
+      if (projectLocation != null && projectLocation.isPrefixOf(location)) {
+        int segmentsToRemove = projectLocation.segmentCount();
+        if (segmentsToRemove > resultProjectPathSegments) {
+          IPath path = project.getFullPath().append(location.removeFirstSegments(segmentsToRemove));
+          IFile resource = resourceFor(path);
+          if (resource != null && !((Resource) resource).isFiltered()) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS134)